Lib/test/test_email/test_email.py

```diff
@@ -728,6 +728,29 @@ def test_nonascii_add_header_with_tspecial(self):
             "attachment; filename*=utf-8''Fu%C3%9Fballer%20%5Bfilename%5D.ppt",
             msg['Content-Disposition'])
 
+    def test_invalid_headers(self):
+        invalid_headers = [
+            ('Invalid Header', 'contains space'),
+            ('Tab\tHeader', 'contains tab'),
+            ('Colon:Header', 'contains colon'),
+            ('', 'Empty name'),
+            (' LeadingSpace', 'starts with space'),
+            ('TrailingSpace ', 'ends with space'),
+        ]
+        for name, value in invalid_headers:
+            with self.assertRaises(ValueError) as cm:
+               Message().add_header(name, value)
+            self.assertIn("Invalid header field name", str(cm.exception))
+
+        invalid_headers = [
+            ('Header\x7F', 'Non-ASCII character'),
+            ('Header\x1F', 'control character'),
+        ]
+        for name, value in invalid_headers:
+            with self.assertRaises(ValueError) as cm:
+               Message().add_header(name, value)
+            self.assertIn(f"Header field name {name!r} contains invalid characters", str(cm.exception))
+
```

Comment:
Unfortunately this only tests your change to add_header, and does not actually test the underlying problem.  If you implement test methods that test `msg['test header'] = 'foo'`, you will find that the invalid field name will be accepted.  add_header is not used in the `__setitem__` path.

The correct location for the fix is in the two header_store_parse methods in policy.py and _policybase.py.  I'd put a helper method in _policybase and call it from those two methods.  I would also consider using a regex to implement the check...I *think* that will be faster, but you might want to test it.

Also, the tests should test both Message (old API) and EmailMessage (new API).  Since add_header uses the `__setitem__` syntax in its implementation you don't actually need to test it directly, but it wouldn't be bad to keep the tests to make sure that path doesn't get broken in the future.  More tests are good unless they are slow ;)  But not required.

Reply:
Tests still need to be converted to test `__setitem__` and both `Message` and `EmailMessage`, but I assume you were waiting for that until I approved the code changes.

Reply:
We have a four way combinatoric here: `__setitem__`, `add_header`, `policy=default`, and `policy=compat32`.  We can get clever and test them all in one nested set of loops:
```
for thispolicy in (policy.default, policy.compat32):
    for method in ('__setitem__', add_header):
        for problem, value in invalid_headers:
            with self.subTest(
                    name=name, 
                    problem=problem,
                    policy=thispolicy.__class__.__name__,
                    method=method,
                    ):
                with self.assertRaises(ValueError) as cm:
                    gatattr( EmailMessage(policy=thispolicy), method(
                        name,
                        problem,
                        )
                ...
```